[LBSE] Fix visual overflow computation

nikolaszimmermann opened the pull request:
#### 6c8fe3a7b838afa10d1655859e8515e8ed034fa7
<pre>
[LBSE] Fix visual overflow computation
<a href="https://bugs.webkit.org/show_bug.cgi?id=262694">https://bugs.webkit.org/show_bug.cgi?id=262694</a>

Reviewed by Rob Buis.

Fix computation of visual overflow rect in SVGBoundingBoxComputation.
Stop taking transformations into account -- the layout related geometry excludes
transformations (CSS/HTML transformations do not influence layout, only paint).

LBSE was changed in 2022 to follow that (<a href="https://trac.webkit.org/changeset/293504/webkit)">https://trac.webkit.org/changeset/293504/webkit)</a>
but we failed to adapt visual overflow handling after this change.

Fixing computeVisualOverflowRect() resolve the issue in general, but broke &lt;foreignObject&gt;
repainting in LBSE. This only uncovered that overflow handling in general is broken for
&lt;foreignObject&gt; (and has always been in WebKit since its introduction in 2010). Stop handling
x/y via transformations, but instead move the block itself along x/y direction.

This change therefore fixes both visual overflow in general, and specifically for &lt;foreignObject&gt;.

Covered by existing tests (svg/overflow/overflow-on-foreignObject.svg fixed in LBSE only).

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/coords-viewattr-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-03-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/foreign-object-skew-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/foreignObject-crash-on-hover-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-1-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-2-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-4-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-5-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-6-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-non-svg-namespaced-element-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/body-background-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/multiple-foreign-objects-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/svg-document-as-direct-child-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/hixie/mixed/008-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/overflow/overflow-on-foreignObject-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/foreignObject-repaint-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/foreignObject-text-clipping-bug-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/text-rescale-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-coords-viewattr-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-foreignObject-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-coords-viewattr-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-foreignObject-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-hixie-mixed-008-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-pixel/svg/overflow/overflow-on-foreignObject-expected.png:
* LayoutTests/svg/overflow/overflow-on-foreignObject.svg:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
(WebCore::RenderSVGForeignObject::layout):
(WebCore::RenderSVGForeignObject::overflowClipRect const):
(WebCore::RenderSVGForeignObject::applyTransform const):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.h:
(WebCore::SVGBoundingBoxComputation::computeVisualOverflowRect):

Canonical link: <a href="https://commits.webkit.org/268981@main">https://commits.webkit.org/268981@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f81a9b4004c068523a8d236a6555e05476acf1c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21429 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22116 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22931 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19585 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21289 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21615 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20840 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21273 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18256 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23785 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19084 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25363 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19248 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19300 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23292 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19832 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16855 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19102 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5084 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23404 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19677 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
```diff
@@ -145,7 +137,7 @@ void RenderSVGForeignObject::updateFromStyle()
 void RenderSVGForeignObject::applyTransform(TransformationMatrix& transform, const RenderStyle& style, const FloatRect& boundingBox, OptionSet<RenderStyle::TransformOperationOption> options) const
 {
-    applySVGTransform(transform, foreignObjectElement(), style, boundingBox, std::nullopt, m_supplementalLayerTransform.isIdentity() ? std::nullopt : std::make_optional(m_supplementalLayerTransform), options);
+    applySVGTransform(transform, foreignObjectElement(), style, boundingBox, std::nullopt, std::nullopt, options);
```
We may want to re-order parameters for convenient default values so we do not need to pass std:nullopt twice.

nikolaszimmermann replied:
Good point - next cleanup PR will include that.

rwlbuis reviewed (approved):
resolve -> resolves

LGTM.

nikolaszimmermann committed:
[LBSE] Fix visual overflow computation
https://bugs.webkit.org/show_bug.cgi?id=262694

Reviewed by Rob Buis.

Fix computation of visual overflow rect in SVGBoundingBoxComputation.
Stop taking transformations into account -- the layout related geometry excludes
transformations (CSS/HTML transformations do not influence layout, only paint).

LBSE was changed in 2022 to follow that (https://trac.webkit.org/changeset/293504/webkit)
but we failed to adapt visual overflow handling after this change.

Fixing computeVisualOverflowRect() resolve the issue in general, but broke <foreignObject>
repainting in LBSE. This only uncovered that overflow handling in general is broken for
<foreignObject> (and has always been in WebKit since its introduction in 2010). Stop handling
x/y via transformations, but instead move the block itself along x/y direction.

This change therefore fixes both visual overflow in general, and specifically for <foreignObject>.

Covered by existing tests (svg/overflow/overflow-on-foreignObject.svg fixed in LBSE only).

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/coords-viewattr-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-03-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/foreign-object-skew-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/foreignObject-crash-on-hover-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-1-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-2-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-4-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-5-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-disallowed-foreign-object-6-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-non-svg-namespaced-element-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/body-background-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/multiple-foreign-objects-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/svg-document-as-direct-child-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/hixie/mixed/008-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/overflow/overflow-on-foreignObject-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/foreignObject-repaint-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/foreignObject-text-clipping-bug-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/text-rescale-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-coords-viewattr-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-foreignObject-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-coords-viewattr-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-foreignObject-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-hixie-mixed-008-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-pixel/svg/overflow/overflow-on-foreignObject-expected.png:
* LayoutTests/svg/overflow/overflow-on-foreignObject.svg:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
(WebCore::RenderSVGForeignObject::layout):
(WebCore::RenderSVGForeignObject::overflowClipRect const):
(WebCore::RenderSVGForeignObject::applyTransform const):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.h:
(WebCore::SVGBoundingBoxComputation::computeVisualOverflowRect):

Canonical link: https://commits.webkit.org/268981@main
